tree/ntuple/v7/src/RNTupleIndex.cxx

```diff
@@ -0,0 +1,137 @@
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include <ROOT/RNTupleIndex.hxx>
+
+namespace {
+ROOT::Experimental::Internal::RNTupleIndex::NTupleIndexValue_t
+CastValuePtr(void *valuePtr, const ROOT::Experimental::RFieldBase &field)
+{
+   ROOT::Experimental::Internal::RNTupleIndex::NTupleIndexValue_t value;
+
+   switch (field.GetValueSize()) {
+   case 1: value = *reinterpret_cast<std::uint8_t *>(valuePtr); break;
+   case 2: value = *reinterpret_cast<std::uint16_t *>(valuePtr); break;
+   case 4: value = *reinterpret_cast<std::uint32_t *>(valuePtr); break;
+   case 8: value = *reinterpret_cast<std::uint64_t *>(valuePtr); break;
+   default: throw ROOT::Experimental::RException(R__FAIL("value size not supported"));
+   }
+
+   return value;
+}
+} // anonymous namespace
+
+ROOT::Experimental::Internal::RNTupleIndex::RNTupleIndex(const std::vector<std::string> &fieldNames,
+                                                         const RPageSource &pageSource)
+   : fPageSource(pageSource.Clone())
+{
+   fPageSource->Attach();
+   auto desc = fPageSource->GetSharedDescriptorGuard();
+
+   fIndexFields.reserve(fieldNames.size());
+
+   for (const auto &fieldName : fieldNames) {
+      auto fieldId = desc->FindFieldId(fieldName);
+      if (fieldId == kInvalidDescriptorId)
+         throw RException(R__FAIL("Could not find field \"" + std::string(fieldName) + "."));
+
+      const auto &fieldDesc = desc->GetFieldDescriptor(fieldId);
+      auto field = fieldDesc.CreateField(desc.GetRef());
+
+      CallConnectPageSourceOnField(*field, *fPageSource);
+
+      fIndexFields.push_back(std::move(field));
+   }
+}
+
+void ROOT::Experimental::Internal::RNTupleIndex::EnsureBuilt() const
+{
+   if (!fIsBuilt)
+      throw RException(R__FAIL("Index has not been built yet"));
+}
+
+std::unique_ptr<ROOT::Experimental::Internal::RNTupleIndex>
+ROOT::Experimental::Internal::RNTupleIndex::Create(const std::vector<std::string> &fieldNames,
+                                                   const RPageSource &pageSource, bool deferBuild)
+{
+   auto index = std::unique_ptr<RNTupleIndex>(new RNTupleIndex(fieldNames, pageSource));
+
+   if (!deferBuild)
+      index->Build();
+
+   return index;
+}
+
+void ROOT::Experimental::Internal::RNTupleIndex::Build()
+{
+   std::vector<RFieldBase::RValue> fieldValues;
+   fieldValues.reserve(fIndexFields.size());
+   for (const auto &field : fIndexFields) {
+      if (!field->IsSimple() || field->GetTypeName() == "float" || field->GetTypeName() == "double") {
+         throw RException(R__FAIL("Cannot use field \"" + field->GetFieldName() + "\" with type \"" +
+                                  field->GetTypeName() + "\" for indexing. Only integral types are allowed."));
+      }
```

Comment:
Perhaps better to explicitly enumerate the supported types, e.g. check `field->GetTypeName()` against a static `std::unordered_set<std::string>`. Or use a dynamic cast with the recent `RIntegralField`.